src/hotspot/share/opto/castnode.hpp

```diff
@@ -32,30 +32,105 @@
+    }
+
+    bool narrows_type() const {
+      return _narrows_type;
+    }
+    void dump_on(outputStream *st) const {
+      st->print("%s", _desc);
+    }
+
+    uint hash() const {
+      return (_floating ? 1 : 0) + (_narrows_type ? 2 : 0);
+    }
+
+    bool cmp(const DependencyType& other) const {
+      return _floating == other._floating && _narrows_type == other._narrows_type;
+    }
+
+    const DependencyType& widen_type_dependency() const {
+      if (_floating) {
+        return FloatingNonNarrowingDependency;
+      }
+      return NonFloatingNonNarrowingDependency;
+    }
+
+    const DependencyType& pinned_dependency() const {
+      if (_narrows_type) {
+        return NonFloatingNarrowingDependency;
+      }
+      return NonFloatingNonNarrowingDependency;
+    }
+
+  private:
+    const bool _floating; // Does this Cast depends on its control input or is it pinned?
+    const bool _narrows_type; // Does this Cast narrows the type i.e. if input type is narrower can it be removed?
+    const char* _desc;
   };
 
-  protected:
-  const DependencyType _dependency;
+public:
+
+  // All the possible combinations of floating/narrowing. Example use cases for each:
+  // FloatingNarrowingDependency is used for range checks: the range check CastII is dependent on the range check and if
+  // its input's type is narrower than the type of the range check, it's safe to be removed.
+  // NonFloatingNonNarrowingDependency is used when a floating node is sunk out of loop: we don't want the cast that
+  // forces the node to be out of loop to be removed in any case
+  // NonFloatingNarrowingDependency is used when an array access is no longer dependent on a single range check (range
+  // check smearing for instance)
+  // FloatingNonNarrowingDependency is used after loop opts when Cast nodes' types are widen so Casts that only differ
+  // by slightly different types can common. Given the type carried by the Cast is no longer accurate, removing a Cast
+  // because its input has a narrower type causes the dependency carried by the Cast to be lost
+  static const DependencyType FloatingNarrowingDependency;
+  static const DependencyType FloatingNonNarrowingDependency;
+  static const DependencyType NonFloatingNarrowingDependency;
+  static const DependencyType NonFloatingNonNarrowingDependency;
+
+protected:
+  const DependencyType& _dependency;
   virtual bool cmp( const Node &n ) const;
   virtual uint size_of() const;
   virtual uint hash() const;    // Check the type
-  const Type* widen_type(const PhaseGVN* phase, const Type* res, BasicType bt) const;
-  Node* find_or_make_integer_cast(PhaseIterGVN* igvn, Node* parent, const TypeInteger* type) const;
+  const TypeInteger* widen_type(const PhaseGVN* phase, const Type* res, BasicType bt) const;
+
+  virtual ConstraintCastNode* make_with(Node* parent, const TypeInteger* type, const DependencyType& dependency) const {
+    ShouldNotReachHere();
+    return nullptr;
+  }
```

Comment:
This always smells like a messed up class hierarchy, when I see default methods with "not implemented". But maybe we can't do much better, and I've done similar things recently 🙈 . A short code comment could be helpful though.

```suggestion
  virtual ConstraintCastNode* make_with(Node* parent, const TypeInteger* type, const DependencyType& dependency) const {
    ShouldNotReachHere(); // Only implemented for CastII and CastLL
    return nullptr;
  }
```